probe-rs/src/architecture/riscv/mod.rs

```diff
@@ -122,86 +149,25 @@ impl<'probe> CoreInterface for Riscv32<'probe> {
     }
 
     fn run(&mut self) -> Result<(), crate::Error> {
-        // TODO: test if core halted?
-
-        // set resume request
-        let mut dmcontrol = Dmcontrol(0);
-        dmcontrol.set_dmactive(true);
-        dmcontrol.set_resumereq(true);
-
-        self.interface.write_dm_register(dmcontrol)?;
-
-        // check if request has been acknowleged
-        let status: Dmstatus = self.interface.read_dm_register()?;
-
-        if !status.allresumeack() {
-            return Err(RiscvError::RequestNotAcknowledged.into());
-        };
-
-        // clear resume request
-        let mut dmcontrol = Dmcontrol(0);
-        dmcontrol.set_dmactive(true);
+        // Before we run, we always perform a single instruction step, to account for possible breakpoints that might get us stuck on the current instruction.
+        self.step()?;
 
-        self.interface.write_dm_register(dmcontrol)?;
+        // resume the core.
+        resume_core(self)?;
 
         Ok(())
     }
 
     fn reset(&mut self) -> Result<(), crate::Error> {
-        log::debug!("Resetting core, setting hartreset bit");
-
-        let mut dmcontrol = Dmcontrol(0);
-        dmcontrol.set_dmactive(true);
-        dmcontrol.set_hartreset(true);
-
-        self.interface.write_dm_register(dmcontrol)?;
-
-        // Read back register to verify reset is supported
-        let readback: Dmcontrol = self.interface.read_dm_register()?;
-
-        if readback.hartreset() {
```

Comment:
Where did those reset bits go?

Reply:
The are in the `reset_and_halt()` method that is being called here.